typing/uniqueness_analysis.ml

```diff
@@ -1058,117 +1062,133 @@ let rec pattern_match_tuple pat values =
     let ext, uf' = pattern_match_single pat paths in
     ext, UF.seq uf uf'
 
-and pattern_match_single pat paths : Ienv.Extension.t * UF.t =
+(** This function ensures the soundness of pattern-matching in the presence
+    of destructive updates on the memory that was matched on.
+    If the pattern-match reads from the underlying memory, we need to ensure
+    either that the memory access is not pushed down or that no destructive
+    updates can be performed on the memory.
+
+    Each pattern falls into one of three cases:
+    - If we do not read from the underlying memory,
+      we do not have to take an action.
+    - We can allow destructive updates later on by borrowing the
+      memory address. Then we have to protect the read from getting
+      pushed down using a unique barrier.
+    - We can disallow any destructive updates following the read
+      by consuming the memory address as aliased. *)
+and pattern_match_barrier pat paths : UF.t =
   let loc = pat.pat_loc in
   let occ = Occurrence.mk loc in
-  (* To read from the allocation, we need to borrow its memory cell
-     and set the unique_barrier. However, we do not read in every case,
-     since the user might want use a wildcard for already-consumed data. *)
-  let no_borrow_memory_address () =
-    Unique_barrier.enable pat.pat_unique_barrier;
-    ignore (Unique_barrier.resolve pat.pat_unique_barrier)
+  Unique_barrier.enable pat.pat_unique_barrier;
+  let no_memory_access () =
+    ignore (Unique_barrier.resolve pat.pat_unique_barrier);
+    UF.unused
   in
   let borrow_memory_address () =
-    Unique_barrier.enable pat.pat_unique_barrier;
     Paths.mark_implicit_borrow_memory_address occ (Read pat.pat_unique_barrier)
       paths
   in
+  let consume_memory_address reason =
+    ignore (Unique_barrier.resolve pat.pat_unique_barrier);
+    Paths.mark_aliased occ reason paths
+  in
   match pat.pat_desc with
-  | Tpat_or (pat0, pat1, _) ->
-    no_borrow_memory_address ();
-    let ext0, uf0 = pattern_match_single pat0 paths in
-    let ext1, uf1 = pattern_match_single pat1 paths in
-    Ienv.Extension.disjunct ext0 ext1, UF.choose uf0 uf1
-  | Tpat_any ->
-    no_borrow_memory_address ();
-    Ienv.Extension.empty, UF.unused
-  | Tpat_var (id, _, _, _) ->
-    no_borrow_memory_address ();
-    Ienv.Extension.singleton id paths, UF.unused
-  | Tpat_alias (pat', id, _, _, _) ->
-    no_borrow_memory_address ();
-    let ext0 = Ienv.Extension.singleton id paths in
-    let ext1, uf = pattern_match_single pat' paths in
-    Ienv.Extension.conjunct ext0 ext1, uf
+  | Tpat_or _ -> no_memory_access ()
+  | Tpat_any -> no_memory_access ()
+  | Tpat_var _ -> no_memory_access ()
```

Comment:
Why is this case correct? Binding a variable looks like a memory access. I'm having a hard time understanding what to do here. Maybe we're answering the question "Is a read of memory required to determine whether the match succeeds?" That sounds plausible. And then it's pretty clear that the answer is "no" here but "yes" in the constant case. But mostly I'm not sure.

Reply:
Your heuristic is exactly right: Binding a variable does not require dereferencing the pointer. That is why there is no memory access required here.